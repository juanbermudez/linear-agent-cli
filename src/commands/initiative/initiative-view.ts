import { Command } from "@cliffy/command"
import { viewInitiative } from "../../utils/linear.ts"
import { formatDate, formatRelativeTime } from "../../utils/display.ts"
import { bold } from "@std/fmt/colors"
import { header, muted } from "../../utils/styling.ts"
import { error as errorColor } from "../../utils/styling.ts"

interface ViewOptions {
  json?: boolean
  format?: string
}

export const viewCommand = new Command()
  .name("view")
  .description("View initiative details")
  .arguments("<initiativeId:string>")
  .option("--no-color", "Disable colored output")
  .option("-j, --json", "Output result as JSON")
  .option("--format <format:string>", "Output format: text|json")
  .action(async (options: ViewOptions, initiativeId: string) => {
    const useJson = options.json || options.format === "json"

    const { Spinner } = await import("@std/cli/unstable-spinner")
    const showSpinner = !useJson && Deno.stdout.isTerminal()
    const spinner = showSpinner
      ? new Spinner({ message: "Fetching initiative..." })
      : null
    spinner?.start()

    try {
      const initiative = await viewInitiative(initiativeId)
      spinner?.stop()

      if (useJson) {
        console.log(
          JSON.stringify(
            {
              initiative: {
                id: initiative.id,
                name: initiative.name,
                slugId: initiative.slugId,
                url: initiative.url,
                description: initiative.description,
                status: initiative.status,
                owner: initiative.owner
                  ? {
                    id: initiative.owner.id,
                    name: initiative.owner.displayName,
                  }
                  : null,
                targetDate: initiative.targetDate,
                createdAt: initiative.createdAt,
                updatedAt: initiative.updatedAt,
                projects: initiative.projects.nodes.map((p) => ({
                  id: p.id,
                  name: p.name,
                  slugId: p.slugId,
                  url: p.url,
                })),
              },
            },
            null,
            2,
          ),
        )
        return
      }

      // Text output
      console.log(header(initiative.name))
      console.log()

      console.log(bold("ID: ") + initiative.slugId)
      console.log(bold("URL: ") + initiative.url)

      if (initiative.description) {
        console.log()
        console.log(bold("Description"))
        console.log(initiative.description)
      }

      console.log()
      console.log(
        bold("Status: ") +
          (initiative.status || "None"),
      )
      console.log(
        bold("Owner: ") +
          (initiative.owner ? initiative.owner.displayName : "Unassigned"),
      )
      console.log(
        bold("Target Date: ") +
          (initiative.targetDate ? formatDate(initiative.targetDate) : "None"),
      )

      console.log()
      console.log(
        bold("Created: ") + formatRelativeTime(initiative.createdAt),
      )
      console.log(
        bold("Updated: ") + formatRelativeTime(initiative.updatedAt),
      )

      if (initiative.projects.nodes.length > 0) {
        console.log()
        console.log(bold(`Projects (${initiative.projects.nodes.length})`))
        for (const project of initiative.projects.nodes) {
          console.log(`  â€¢ ${project.name} (${project.slugId})`)
          console.log(`    ${muted(project.url)}`)
        }
      }
    } catch (err) {
      spinner?.stop()
      const errorMsg = (err as Error).message.includes("not found")
        ? `Initiative '${initiativeId}' not found`
        : `Failed to fetch initiative: ${(err as Error).message}`
      if (useJson) {
        console.error(
          JSON.stringify(
            {
              success: false,
              error: {
                code: (err as Error).message.includes("not found")
                  ? "NOT_FOUND"
                  : "API_ERROR",
                message: errorMsg,
              },
            },
            null,
            2,
          ),
        )
      } else {
        console.error(errorColor(`Error: ${errorMsg}`))
      }
      Deno.exit(1)
    }
  })
