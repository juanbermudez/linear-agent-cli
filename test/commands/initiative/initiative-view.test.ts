import { snapshotTest } from "@cliffy/testing"
import { viewCommand } from "../../../src/commands/initiative/initiative-view.ts"
import {
  commonDenoArgs,
  setupMockLinearServer,
} from "../../utils/test-helpers.ts"

// Test help output
await snapshotTest({
  name: "Initiative View Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs: commonDenoArgs,
  async fn() {
    await viewCommand.parse()
  },
})

// Test viewing an initiative with details
await snapshotTest({
  name: "Initiative View Command - Full Details",
  meta: import.meta,
  colors: false,
  args: ["init-123", "--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "ViewInitiative",
        variables: { id: "init-123" },
        response: {
          data: {
            initiative: {
              id: "init-123",
              name: "Q1 2025 Platform Goals",
              slugId: "q1-2025-platform-goals",
              url: "https://linear.app/test/initiative/q1-2025-platform-goals",
              description:
                "Focus on platform stability and performance improvements",
              status: {
                id: "status-active",
                name: "Active",
                type: "started",
              },
              owner: {
                name: "alice",
                displayName: "Alice Smith",
              },
              targetDate: "2025-03-31",
              projects: {
                nodes: [
                  {
                    id: "project-1",
                    name: "API Redesign",
                    slugId: "api-redesign",
                    status: {
                      name: "In Progress",
                      type: "started",
                    },
                  },
                  {
                    id: "project-2",
                    name: "Database Migration",
                    slugId: "database-migration",
                    status: {
                      name: "Planned",
                      type: "planned",
                    },
                  },
                ],
              },
              createdAt: "2024-01-01T10:00:00Z",
              updatedAt: "2024-01-15T14:30:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test viewing with JSON output
await snapshotTest({
  name: "Initiative View Command - JSON Output",
  meta: import.meta,
  colors: false,
  args: ["init-456", "--json", "--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "ViewInitiative",
        variables: { id: "init-456" },
        response: {
          data: {
            initiative: {
              id: "init-456",
              name: "API Modernization",
              slugId: "api-modernization",
              url: "https://linear.app/test/initiative/api-modernization",
              description: "Modernize our API infrastructure",
              status: {
                id: "status-planned",
                name: "Planned",
                type: "planned",
              },
              owner: {
                name: "bob",
                displayName: "Bob Jones",
              },
              targetDate: "2025-06-30",
              projects: {
                nodes: [],
              },
              createdAt: "2024-01-05T09:00:00Z",
              updatedAt: "2024-01-10T11:20:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test viewing initiative without optional fields
await snapshotTest({
  name: "Initiative View Command - Minimal Fields",
  meta: import.meta,
  colors: false,
  args: ["init-minimal", "--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "ViewInitiative",
        variables: { id: "init-minimal" },
        response: {
          data: {
            initiative: {
              id: "init-minimal",
              name: "Simple Initiative",
              slugId: "simple-initiative",
              url: "https://linear.app/test/initiative/simple-initiative",
              description: null,
              status: null,
              owner: null,
              targetDate: null,
              projects: {
                nodes: [],
              },
              createdAt: "2024-01-10T15:00:00Z",
              updatedAt: "2024-01-10T15:00:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test viewing initiative with many projects
await snapshotTest({
  name: "Initiative View Command - With Multiple Projects",
  meta: import.meta,
  colors: false,
  args: ["init-many-projects", "--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "ViewInitiative",
        variables: { id: "init-many-projects" },
        response: {
          data: {
            initiative: {
              id: "init-many-projects",
              name: "Platform Modernization",
              slugId: "platform-modernization",
              url: "https://linear.app/test/initiative/platform-modernization",
              description: "Comprehensive platform upgrade",
              status: {
                id: "status-active",
                name: "Active",
                type: "started",
              },
              owner: {
                name: "alice",
                displayName: "Alice Smith",
              },
              targetDate: "2025-12-31",
              projects: {
                nodes: [
                  {
                    id: "project-1",
                    name: "API Redesign",
                    slugId: "api-redesign",
                    status: { name: "In Progress", type: "started" },
                  },
                  {
                    id: "project-2",
                    name: "Database Migration",
                    slugId: "database-migration",
                    status: { name: "Completed", type: "completed" },
                  },
                  {
                    id: "project-3",
                    name: "Frontend Upgrade",
                    slugId: "frontend-upgrade",
                    status: { name: "In Progress", type: "started" },
                  },
                  {
                    id: "project-4",
                    name: "Security Audit",
                    slugId: "security-audit",
                    status: { name: "Planned", type: "planned" },
                  },
                ],
              },
              createdAt: "2024-01-01T10:00:00Z",
              updatedAt: "2024-01-15T14:30:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test viewing by slug ID
await snapshotTest({
  name: "Initiative View Command - By Slug ID",
  meta: import.meta,
  colors: false,
  args: ["q1-2025-goals", "--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "ViewInitiative",
        variables: { id: "q1-2025-goals" },
        response: {
          data: {
            initiative: {
              id: "init-789",
              name: "Q1 2025 Goals",
              slugId: "q1-2025-goals",
              url: "https://linear.app/test/initiative/q1-2025-goals",
              description: "Quarterly goals for Q1",
              status: {
                id: "status-active",
                name: "Active",
                type: "started",
              },
              owner: {
                name: "alice",
                displayName: "Alice Smith",
              },
              targetDate: "2025-03-31",
              projects: {
                nodes: [],
              },
              createdAt: "2024-01-01T10:00:00Z",
              updatedAt: "2024-01-15T14:30:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})
