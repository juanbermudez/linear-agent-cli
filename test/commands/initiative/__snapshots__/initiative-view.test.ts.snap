export const snapshot = {};

snapshot[`Initiative View Command - Help Text 1`] = `
stdout:
"
Usage: view <initiativeId>

Description:

  View initiative details

Options:

  -h, --help            - Show this help.           
  --no-color            - Disable colored output    
  -j, --json            - Output result as JSON     
  --format    <format>  - Output format: text|json  

"
stderr:
""
`;
